Lecture4/Task4/Task4L4.java

```diff
@@ -0,0 +1,308 @@
+        } else {
+            listOfSelection = listOfSelection.stream()
+                                             .filter(each -> !each.getTransfer().equalsIgnoreCase("no"))
+                                             .collect(Collectors.toList());
+        }
+    }
+
+    private static boolean isUserNeedTransfer() {
+        String transfer = null;
+        while (transfer == null) {
+            System.out.print("Do you need transfer? (y/n) ");
+            Scanner input = new Scanner(System.in);
+            String userInput;
+            userInput = input.nextLine();
+            try {
+                if (userInput.matches(REGEX_YES_NO)) {
+                    transfer = userInput;
+                } else {
+                    throw new YesNoAnswerException(MESSAGE_TRY_AGAIN);
+                }
+            } catch (YesNoAnswerException yesNoAnswerEx) {
+                System.out.print(yesNoAnswerEx.getMessage());
+            }
+        }
+        return transfer.equalsIgnoreCase("y");
+    }
+
+    private static void selectVoucher() {
+        int likeVoucherId = 0;
+        while (likeVoucherId == 0) {
+            System.out.print("Lets select a voucher you like (by ID): ");
+            Scanner input = new Scanner(System.in);
+            int userInput;
+            try {
+                userInput = input.nextInt();
+            } catch (InputMismatchException inputMismatchEx) {
+                System.out.print(MESSAGE_TRY_AGAIN);
+                continue;
+            }
+            try {
+                if (userInput >= 1 && userInput <= listOfSelection.size()) {
+                    likeVoucherId = userInput;
+                } else {
+                    throw new VoucherNotFoundException(MESSAGE_TRY_AGAIN);
+                }
+            } catch (VoucherNotFoundException voucherNotFoundEx) {
+                System.out.print(voucherNotFoundEx.getMessage());
+            }
+        }
+        System.out.print("\nYou've selected:\n" + "ID: " + (likeVoucherId) + " | " + listOfSelection.get(likeVoucherId - 1));
+        writeSelectedToFile(listOfSelection.get(likeVoucherId - 1));
+    }
+
+    private static void provideFinalOffer() {
+        try {
+            if (!listOfSelection.isEmpty()) {
+                for (int i = 0; i < listOfSelection.size(); i++) {
+                    System.out.print("ID: " + (i + 1) + " | " + listOfSelection.get(i));
+                }
+                selectVoucher();
+            } else {
+                throw new EmptyListException(MESSAGE_SORRY);
+            }
+        } catch (EmptyListException emptyListEx) {
+            emptyListEx.getMessageAndExit();
+        }
+    }
+
+    private static void writeSelectedToFile(Voucher selectedVoucher) {
+        String[] voucherDetails = selectedVoucher.toString().split(" \\| ");
+        try {
+            FileWriter voucherDetailsWriter = new FileWriter("Your voucher.txt");
```

Comment:
in additional to writing voucher as string, implement a method to write voucher as object

Reply:
fixed in Tasks1-4-fixed package